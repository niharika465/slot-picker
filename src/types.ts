export type SlotTime = { displayTime: string }[];

export type DateInfo = {
  day: string;
  weekday: string;
  date: string;
};
