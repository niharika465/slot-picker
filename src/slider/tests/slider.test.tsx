import { render, screen, fireEvent } from '@testing-library/react';
import Slider from '../slider';
import { DateInfo } from 'types';

// Mock class for IntersectionObserver
class IntersectionObserverMock {
  private callback: IntersectionObserverCallback;
  private observedElements: Element[] = [];

  constructor(callback: IntersectionObserverCallback) {
    this.callback = callback;
  }

  observe(element: Element) {
    this.observedElements.push(element);
    // Trigger the callback
    this.callback(
      [
        {
          isIntersecting: true,
          boundingClientRect: element.getBoundingClientRect(),
          intersectionRatio: 1,
          target: element,
          time: 0,
        } as IntersectionObserverEntry,
      ],
      this // Pass the mock instance as the second argument
    );
  }

  unobserve(element: Element) {
    this.observedElements = this.observedElements.filter((e) => e !== element);
  }

  disconnect() {
    this.observedElements = [];
  }

  // Added these properties incase typescript doesn't recognize them
  get root() {
    return null;
  }

  get rootMargin() {
    return '';
  }

  get thresholds() {
    return [];
  }

  takeRecords() {
    return [];
  }
}

// global scope for the observer
// eslint-disable-next-line @typescript-eslint/no-explicit-any
(global as any).IntersectionObserver = IntersectionObserverMock;

const mockCards: DateInfo[] = [
  { day: '30', weekday: 'Mon', date: '2024-09-30' },
  { day: '01', weekday: 'Tue', date: '2024-10-01' },
  { day: '02', weekday: 'Wed', date: '2024-10-02' },
  { day: '03', weekday: 'Thu', date: '2024-09-03' },
  { day: '04', weekday: 'Fri', date: '2024-10-04' },
  { day: '05', weekday: 'Sat', date: '2024-10-05' },
];

describe('Slider', () => {
  const setSelectedSlot = jest.fn();

  test('renders cards', () => {
    const { asFragment } = render(
      <Slider
        cards={mockCards}
        setSelectedSlot={setSelectedSlot}
        heading="Pick a date"
        selectedSlot=""
      />
    );
    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByText('Pick a date')).toBeInTheDocument();
    expect(screen.getByText('Mon')).toBeInTheDocument();
    expect(screen.getByText('Tue')).toBeInTheDocument();
    expect(screen.getByText('Wed')).toBeInTheDocument();
  });

  test('renders navigation buttons', () => {
    render(
      <Slider
        cards={mockCards}
        setSelectedSlot={setSelectedSlot}
        heading="Pick a date"
        selectedSlot=""
      />
    );

    expect(screen.getByTestId('ArrowForwardIcon')).toBeInTheDocument();
    expect(screen.getByTestId('ArrowBackIcon')).toBeInTheDocument();
  });

  test('does not navigate back when at the beginning', () => {
    render(
      <Slider
        cards={mockCards}
        setSelectedSlot={setSelectedSlot}
        heading="Pick a date"
        selectedSlot=""
      />
    );

    const backButton = screen
      .getByTestId('ArrowBackIcon')
      .closest('div') as HTMLDivElement;
    fireEvent.click(backButton);

    expect(backButton).toHaveClass('navigation-btn--disabled');
  });

  test('selects a card and calls setSelectedSlot', () => {
    render(
      <Slider
        cards={mockCards}
        setSelectedSlot={setSelectedSlot}
        heading="Pick a date"
        selectedSlot=""
      />
    );

    const card = screen.getByText('Mon');
    fireEvent.click(card);

    expect(setSelectedSlot).toHaveBeenCalledWith('2024-09-30');
  });

  test('highlights the selected card', () => {
    render(
      <Slider
        cards={mockCards}
        setSelectedSlot={setSelectedSlot}
        heading="Pick a date"
        selectedSlot="2024-09-30"
      />
    );

    const selectedCard = screen
      .getByText('Mon')
      .closest('.slide-item') as HTMLDivElement;
    fireEvent.click(selectedCard);
    expect(selectedCard).toHaveClass('slide-item--selected');
  });
});
