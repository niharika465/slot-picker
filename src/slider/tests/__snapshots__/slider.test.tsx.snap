// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider renders cards 1`] = `
<DocumentFragment>
  <div>
    <div
      class="heading"
    >
      Pick a date
    </div>
    <div
      class="slider"
    >
      <div
        class="navigation-btn navigation-btn--disabled"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-tcjmn5-MuiSvgIcon-root"
          data-testid="ArrowBackIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20z"
          />
        </svg>
      </div>
      <div
        class="slide-container"
      >
        <div
          class="slide-item"
        >
          <div
            class=""
          >
            30
          </div>
          <div
            class="slide-item__day"
          >
            Mon
          </div>
        </div>
        <div
          class="slide-item"
        >
          <div
            class=""
          >
            01
          </div>
          <div
            class="slide-item__day"
          >
            Tue
          </div>
        </div>
        <div
          class="slide-item"
        >
          <div
            class=""
          >
            02
          </div>
          <div
            class="slide-item__day"
          >
            Wed
          </div>
        </div>
        <div
          class="slide-item"
        >
          <div
            class=""
          >
            03
          </div>
          <div
            class="slide-item__day"
          >
            Thu
          </div>
        </div>
        <div
          class="slide-item"
        >
          <div
            class=""
          >
            04
          </div>
          <div
            class="slide-item__day"
          >
            Fri
          </div>
        </div>
        <div
          class="slide-item"
        >
          <div
            class=""
          >
            05
          </div>
          <div
            class="slide-item__day"
          >
            Sat
          </div>
        </div>
      </div>
      <div
        class="navigation-btn navigation-btn--disabled"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-tcjmn5-MuiSvgIcon-root"
          data-testid="ArrowForwardIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="m12 4-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8z"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
